### PR TITLE
feat(new schema): implement backfill API to streamline old -> new schema migration

### PR DESCRIPTION
**Pull request by jsdonn**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java [hunk 1]
```diff
@@ -362,6 +362,21 @@ public <ASPECT extends RecordTemplate> void updateLocalIndex(@Nonnull URN urn, @
...
+          + "parameter is properly set to NEW_SCHEMA_ONLY or DUAL_SCHEMA if you wish to use entity tables.");
+    }
+    PrimaryKey key = new PrimaryKey(urn.toString(), ModelUtils.getAspectName(newValue.getClass()), LATEST_VERSION);
+    EbeanMetadataAspect result = _server.find(EbeanMetadataAspect.class, key);
+    if (result == null) {
+      return;
+    }
+    AuditStamp auditStamp = makeAuditStamp(result);
+    _localAccess.add(urn, newValue, (Class<ASPECT>) newValue.getClass(), auditStamp);
+  }
```
Comment (zhixuanjia) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:378]: Just think aloud, backfill process will run together with live ingestion pipeline, meaning there could be MCEs being ingested. Could there be a case such that: `EbeanMetadataAspect result = _server.find(EbeanMetadataAspect.class, key);` find valueA and then ingestion pipeline update entity with latest version valueB, then `_localAccess.add(...)` will change the value back to valueA which is actually incorrect (should be valueB instead). We might have to lock the records before updating them.

Comment (jsdonn) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:378]: Thanks for the heads up, I agree with you that the use case you presented is possible. I will run these methods in a single transaction.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 2]
```diff
@@ -617,6 +617,15 @@ protected abstract <ASPECT extends RecordTemplate> long saveLatest(@Nonnull URN
...
       long version);
 
+  /**
+   * Saves the new value of an aspect to entity tables. This is used when backfilling metadata from the old schema to
+   * the new schema.
+   *
+   * @param urn the URN for the entity the aspect is attached to
+   * @param newValue {@link RecordTemplate} of the new value of aspect
+   */
+  public abstract <ASPECT extends RecordTemplate> void updateEntityTables(@Nonnull URN urn, @Nullable ASPECT newValue);
```
Comment (zhixuanjia) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:627]: how about calling this `latestValue`?

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanLocalDAOTest.java [hunk 3]
```diff
@@ -2879,6 +2880,77 @@ public void testOptimisticLockException() {
         new Timestamp(123));
   }
 
+  @Test
```
Comment (zhixuanjia) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanLocalDAOTest.java:2883]: Did you run these newly added tests? Just want to make sure we maintain `all tests passed` status.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java [hunk 4]
```diff
@@ -362,6 +362,24 @@ public <ASPECT extends RecordTemplate> void updateLocalIndex(@Nonnull URN urn, @
...
 
+  @Override
+  public <ASPECT extends RecordTemplate> void updateEntityTables(@Nonnull URN urn, @Nonnull ASPECT newValue) {
+    if (_schemaConfig == SchemaConfig.OLD_SCHEMA_ONLY) {
+      throw new UnsupportedOperationException("The DAO is set to use the OLD_SCHEMA_ONLY. Please check that the schemaConfig"
+          + "parameter is properly set to NEW_SCHEMA_ONLY or DUAL_SCHEMA if you wish to use entity tables.");
+    }
+    PrimaryKey key = new PrimaryKey(urn.toString(), ModelUtils.getAspectName(newValue.getClass()), LATEST_VERSION);
+    runInTransactionWithRetry(() -> {
+      EbeanMetadataAspect result = _server.find(EbeanMetadataAspect.class, key);
```
Comment (zhixuanjia) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java:373]: how about
```
EbeanMetadataAspect result  = _server.createQuery(EbeanMetadataAspect.class).setId(new PrimaryKey()).forUpdate().findOne();
```
Basically, its doing `select ... for update`;
More info https://dba.stackexchange.com/questions/46459/putting-a-select-statement-in-a-transaction

Comment (jsdonn) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java]: Good idea, thanks. Updated.

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanLocalDAOTest.java (hunk 3 above)
Comment (jsdonn) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanLocalDAOTest.java:2895]: Yes I did.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 2 above)
Comment (jsdonn) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:627]: Thanks for the suggestion, updated.

**Review by zhixuanjia — APPROVED** [review]

**Review by keremsahin1 — COMMENTED** [review]
LGTM, thanks!

**Pull request merged**